File: packages/python/queue.py
class Queue:
    def __init__(self):
        self.items = []
        
    def enqueue(self, item):
        self.items.append(item)
        
    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        return None
        
    def is_empty(self):
        return len(self.items) == 0

class Deque:
    def __init__(self):
        self.items = []
        
    def add_front(self, item):
        self.items.insert(0, item)
        
    def add_rear(self, item):
        self.items.append(item)
        
    def remove_front(self):
        if not self.is_empty():
            return self.items.pop(0)
        return None
        
    def remove_rear(self):
        if not self.is_empty():
            return self.items.pop()
        return None
        
    def is_empty(self):
        return len(self.items) == 0

class CircularQueue:
    def __init__(self, k):
        self.k = k
        self.queue = [None] * k
        self.head = self.tail = -1

    def enqueue(self, data):
        if ((self.tail + 1) % self.k == self.head):
            return False
        elif (self.head == -1):
            self.head = 0
            self.tail = 0
            self.queue[self.tail] = data
        else:
            self.tail = (self.tail + 1) % self.k
            self.queue[self.tail] = data
        return True

    def dequeue(self):
        if (self.head == -1):
            return None
        elif (self.head == self.tail):
            temp = self.queue[self.head]
            self.head = -1
            self.tail = -1
            return temp
        else:
            temp = self.queue[self.head]
            self.head = (self.head + 1) % self.k
            return temp

    def is_empty(self):
        return self.head == -1

File: packages/js/queue.js
class Queue {
    constructor() {
        this.items = [];
    }
    enqueue(element) {
        this.items.push(element);
    }
    dequeue() {
        if (this.isEmpty()) return "Underflow";
        return this.items.shift();
    }
    isEmpty() {
        return this.items.length == 0;
    }
}

class Deque {
    constructor() {
        this.items = [];
    }
    addFront(element) {
        this.items.unshift(element);
    }
    addRear(element) {
        this.items.push(element);
    }
    removeFront() {
        if (this.isEmpty()) return "Underflow";
        return this.items.shift();
    }
    removeRear() {
        if (this.isEmpty()) return "Underflow";
        return this.items.pop();
    }
    isEmpty() {
        return this.items.length == 0;
    }
}

class CircularQueue {
    constructor(k) {
        this.items = new Array(k);
        this.maxSize = k;
        this.head = -1;
        this.tail = -1;
    }
    enqueue(element) {
        if (this.isFull()) return false;
        if (this.isEmpty()) {
            this.head = 0;
        }
        this.tail = (this.tail + 1) % this.maxSize;
        this.items[this.tail] = element;
        return true;
    }
    dequeue() {
        if (this.isEmpty()) return false;
        if (this.head === this.tail) {
            this.head = -1;
            this.tail = -1;
        } else {
            this.head = (this.head + 1) % this.maxSize;
        }
        return true;
    }
    Front() {
        if (this.isEmpty()) return -1;
        return this.items[this.head];
    }
    Rear() {
        if (this.isEmpty()) return -1;
        return this.items[this.tail];
    }
    isEmpty() {
        return this.head === -1;
    }
    isFull() {
        return (this.tail + 1) % this.maxSize === this.head;
    }
}

module.exports = { Queue, Deque, CircularQueue };
